### PR TITLE
Quest search

### PR DESCRIPTION
**Pull request by SKProCH**
Implementation of [this feature request](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/6699)

https://user-images.githubusercontent.com/29896317/103277685-a4b5ad00-49da-11eb-9bae-635f96a670db.mp4

### CODE REVIEW

File: src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                                iQuestLineDBEntry)
+                );
+            }).collect(Collectors.toList());
+        }
+        return questList.iterator();
+    }
+
+    @Override
+    protected void queryMatches(QuestSearchEntry value, String query, ArrayDeque<QuestSearchEntry> results) {
+        if (("" + value.getQuest().getID()).contains(query) ||
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java:42]: use toString here please

File: src/main/java/betterquesting/api2/client/gui/controls/PanelButtonQuest.java [hunk 2]
```diff
@@ -34,45 +34,47 @@
...
 			case REPEATABLE:
-				txFrame = main ? PresetTexture.QUEST_MAIN_4.getTexture() : PresetTexture.QUEST_NORM_4.getTexture();
+				txFrame1 = main ? PresetTexture.QUEST_MAIN_4.getTexture() : PresetTexture.QUEST_NORM_4.getTexture();
 				txIconCol = PresetColor.QUEST_ICON_REPEATABLE.getColor();
 				break;
         }
-        
-        IGuiTexture btnTx = new GuiTextureColored(txFrame, txIconCol);
+
+		txFrame = txFrame1;
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/client/gui/controls/PanelButtonQuest.java:76]: why was this added?

File: src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+public abstract class CanvasQuestSearch extends CanvasSearch<QuestSearchEntry, QuestSearchEntry> {
+    private List<QuestSearchEntry> questList;
+
+    public CanvasQuestSearch(IGuiRect rect) {
+        super(rect);
+    }
+
+    @Override
+    protected Iterator<QuestSearchEntry> getIterator() {
+        if (questList == null){
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java:28]: invert the if here please, 
``
if (questList != null)
    return questList.iterator();
``
then continue your regular code. this reduces nesting

File: src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                        new QuestSearchEntry(new DBEntry<>(iQuestLineEntryDBEntry.getID(), QuestDatabase.INSTANCE.getValue(iQuestLineEntryDBEntry.getID())),
+                                iQuestLineDBEntry)
+                );
+            }).collect(Collectors.toList());
+        }
+        return questList.iterator();
+    }
+
+    @Override
+    protected void queryMatches(QuestSearchEntry value, String query, ArrayDeque<QuestSearchEntry> results) {
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java:41]: `value` should have a more precise name, like searchEntry or simply put entry

File: src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    private List<QuestSearchEntry> questList;
+
+    public CanvasQuestSearch(IGuiRect rect) {
+        super(rect);
+    }
+
+    @Override
+    protected Iterator<QuestSearchEntry> getIterator() {
+        if (questList == null){
+            questList = QuestLineDatabase.INSTANCE.getEntries().stream().flatMap(iQuestLineDBEntry -> {
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java:29]: This function is unreadable in this syntax. Either format it in a more readable fashon or put it in a seperate function and call it here.

File: src/main/java/betterquesting/client/gui2/GuiQuestLines.java [hunk 6]
```diff
@@ -518,4 +526,77 @@ private void refreshClaimAll()
...
+
+        PanelTextField<String> searchBox = new PanelTextField<>(new GuiTransform(GuiAlign.TOP_EDGE, new GuiPadding(0, 16, 8, -32), 0), "", FieldFilterString.INSTANCE);
+        searchBox.setWatermark("Search...");
+        cvRight.addPanel(searchBox);
+
+        GuiQuestLines thisGui = this;
+        UUID playerUUID = QuestingAPI.getQuestingUUID(mc.thePlayer);
+        CanvasQuestSearch canvasQuestSearch = new CanvasQuestSearch(new GuiTransform(GuiAlign.FULL_BOX, new GuiPadding(0, 32, 8, 24), 0)) {
+            @Override
+            protected boolean addResult(QuestSearchEntry entry, int index, int cachedWidth) {
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/client/gui2/GuiQuestLines.java:552]: please do not use inline functions.

File: src/main/java/betterquesting/client/gui2/GuiQuestLines.java [hunk 7]
```diff
@@ -518,4 +526,77 @@ private void refreshClaimAll()
...
+                PanelButtonCustom buttonContainer = new PanelButtonCustom(new GuiRectangle(0, index * 32, cachedWidth, 32, 0), 2);
+                buttonContainer.setCallback(panelButtonCustom -> {
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/client/gui2/GuiQuestLines.java:554]: this could be placed in a seperate method

File: src/main/java/betterquesting/client/gui2/GuiQuestLines.java [hunk 8]
```diff
@@ -518,4 +526,77 @@ private void refreshClaimAll()
...
+                    if (!buttonContainer.isActive()) return;
+                    thisGui.initPanel();
+                    openQuestLine(entry.getQuestLineEntry());
+                    int selectedQuestId = entry.getQuest().getID();
+                    Optional<PanelButtonQuest> targetQuestButton = cvQuest.getQuestButtons().stream().filter(panelButtonQuest -> panelButtonQuest.getStoredValue().getID() == selectedQuestId).findFirst();
+                    targetQuestButton.ifPresent(panelButtonQuest ->{
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/client/gui2/GuiQuestLines.java:560]: this could be placed in a seperate method

File: src/main/java/betterquesting/client/gui2/GuiQuestLines.java [hunk 9]
```diff
@@ -518,4 +526,77 @@ private void refreshClaimAll()
...
+                                        1, 0
+                                ));
+                        panelButtonQuest.setTextures(newTexture, newTexture, newTexture);
+                    });
+                });
+                buttonContainer.setActive(QuestCache.isQuestShown(entry.getQuest().getValue(), playerUUID, mc.thePlayer));
+                this.addPanel(buttonContainer);
+
+                PanelButtonQuest questButton = new PanelButtonQuest(new GuiRectangle(2, 2, 28, 28), 0, "", entry.getQuest());
+                questButton.setCallback(value -> {
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/client/gui2/GuiQuestLines.java:574]: this could be placed in a seperate method

File: src/main/java/betterquesting/client/gui2/GuiQuestLines.java [hunk 10]
```diff
@@ -518,4 +526,77 @@ private void refreshClaimAll()
...
+                    mc.displayGuiScreen(GuiHome.bookmark);
+                });
+                buttonContainer.addPanel(questButton);
+
+                buttonContainer.addPanel(new PanelGeneric(new GuiRectangle(36, 2, 14, 14, 0), new OreDictTexture(1F, entry.getQuestLineEntry().getValue().getProperty(NativeProps.ICON), false, true)));
+
+                PanelTextBox chapterName = new PanelTextBox(new GuiRectangle(56, 6, cachedWidth - 56, 16), QuestTranslation.translate(entry.getQuestLineEntry().getValue().getProperty(NativeProps.NAME)));
+                buttonContainer.addPanel(chapterName);
+
+                PanelTextBox questName = new PanelTextBox(new GuiRectangle(36, 20, cachedWidth - 40, 16), QuestTranslation.translate(entry.getQuest().getValue().getProperty(NativeProps.NAME)));
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/client/gui2/GuiQuestLines.java:588]: All of the PanelTextBox -> .addPanel calls can be placed in a multi-purpose seperate method

File: src/main/java/betterquesting/api2/client/gui/controls/PanelButtonCustom.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+        return (click == 0 || click == 1) && contains;
+    }
+
+    @Override
+    public boolean onMouseRelease(int mx, int my, int click) {
+        boolean released = super.onMouseRelease(mx, my, click);
+        if (released) return true;
+
+        if (pendingRelease) {
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/client/gui/controls/PanelButtonCustom.java:88]: invert this to ``if (pendingRelease) return false``

**Review by bartimaeusnek — CHANGES_REQUESTED** [review]

File: src/main/java/betterquesting/api2/client/gui/controls/PanelButtonQuest.java [hunk 12]
```diff
@@ -34,45 +34,47 @@
 {
     public final GuiRectangle rect;
     public final EntityPlayer player;
-    
-    public PanelButtonQuest(GuiRectangle rect, int id, String txt, DBEntry<IQuest> value)
+	public final IGuiTexture txFrame;
```
Comment (Glease) [inline src/main/java/betterquesting/api2/client/gui/controls/PanelButtonQuest.java:37]: Indentation is off.

File: src/main/java/betterquesting/api2/client/gui/controls/PanelButtonQuest.java (hunk 2 above)
Comment (SKProCH) [reply src/main/java/betterquesting/api2/client/gui/controls/PanelButtonQuest.java:76]: Because txFrame are marked as final. IDEA suggest me to do this

**Review by Glease — CHANGES_REQUESTED** [review]

File: src/main/java/betterquesting/client/gui2/GuiQuestLines.java (hunk 6 above)
Comment (SKProCH) [reply src/main/java/betterquesting/client/gui2/GuiQuestLines.java:552]: What should I do then?

Comment (bartimaeusnek) [reply src/main/java/betterquesting/client/gui2/GuiQuestLines.java:552]: So currently youre using an anonymous class in which you override a single function, it would be cleaner, if you would just make a new class file and use that here, instead of this approach

File: src/main/java/betterquesting/api2/client/gui/controls/PanelButtonQuest.java (hunk 2 above)
Comment (bartimaeusnek) [reply src/main/java/betterquesting/api2/client/gui/controls/PanelButtonQuest.java:76]: I don't think you need to introduce the txFrame1 variable at all, tbh. You can leave this as it was before...

**Comment by Dream-Master** [comment]
@bartimaeusnek you said me in pm more changes are needed.

File: src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java [hunk 13]
```diff
@@ -1,48 +1,106 @@
...
-        if (questList == null){
-            questList = QuestLineDatabase.INSTANCE.getEntries().stream().flatMap(iQuestLineDBEntry -> {
-                IQuestLine value = iQuestLineDBEntry.getValue();
-                return value.getEntries().stream().map(iQuestLineEntryDBEntry ->
-                        new QuestSearchEntry(new DBEntry<>(iQuestLineEntryDBEntry.getID(), QuestDatabase.INSTANCE.getValue(iQuestLineEntryDBEntry.getID())),
-                                iQuestLineDBEntry)
-                );
-            }).collect(Collectors.toList());
-        }
+        if (questList != null) questList.iterator();
```
Comment (Skiftcha) [inline src/main/java/betterquesting/api2/client/gui/panels/lists/CanvasQuestSearch.java:44]: return missing
